src/main/java/org/elasticsearch/gateway/GatewayMetaState.java

```diff
@@ -219,6 +217,22 @@ public void clusterChanged(ClusterChangedEvent event) {
 
             // check and write changes in indices
             for (IndexMetaData indexMetaData : newMetaData) {
+
+                boolean shardsAllocatedOnThisNodeInLastClusterState = true;
+                if (isDataOnlyNode(state)) {
```

Comment:
I think this logic can be simplified a lot if we don't need to deal with deletes? That continue block at the end is confusing to me?